File: threedollar-api-user/src/main/java/com/depromeet/threedollar/api/service/auth/AuthServiceProvider.java
package com.depromeet.threedollar.api.service.auth;

import com.depromeet.threedollar.api.service.auth.policy.AppleAuthService;
import com.depromeet.threedollar.api.service.auth.policy.GoogleAuthService;
import com.depromeet.threedollar.api.service.auth.policy.KaKaoAuthService;
import com.depromeet.threedollar.common.exception.model.InternalServerException;
import com.depromeet.threedollar.domain.user.domain.user.UserSocialType;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import javax.annotation.PostConstruct;
import java.util.HashMap;
import java.util.Map;

@RequiredArgsConstructor
@Component
public class AuthServiceProvider {

    private static final Map<UserSocialType, AuthService> authServiceMap = new HashMap<>();

    private final AppleAuthService appleAuthService;
    private final KaKaoAuthService kaKaoAuthService;
    private final GoogleAuthService googleAuthService;

    @PostConstruct
    void initializeAuthServicesMap() {
        authServiceMap.put(UserSocialType.APPLE, appleAuthService);
        authServiceMap.put(UserSocialType.KAKAO, kaKaoAuthService);
        authServiceMap.put(UserSocialType.GOOGLE, googleAuthService);
        validateInitializeAuthServices();
    }

    private void validateInitializeAuthServices() {
        for (UserSocialType userSocialType : UserSocialType.values()) {
            if (authServiceMap.get(userSocialType) == null) {
                throw new InternalServerException(String.format("모든 소셜에 대한 AuthService가 등록되지 않았습니다. 해당 소셜 타입(%s)의 AuthService를 등록해주세요.", userSocialType));
            }
        }
    }

    public AuthService getAuthService(UserSocialType socialType) {
        return authServiceMap.get(socialType);
    }

}

File: threedollar-api-boss/src/main/kotlin/com/depromeet/threedollar/api/boss/service/auth/AuthServiceProvider.kt
package com.depromeet.threedollar.api.boss.service.auth

import com.depromeet.threedollar.api.boss.service.auth.policy.AppleAuthService
import com.depromeet.threedollar.api.boss.service.auth.policy.GoogleAuthService
import com.depromeet.threedollar.api.boss.service.auth.policy.KaKaoAuthService
import com.depromeet.threedollar.api.boss.service.auth.policy.NaverAuthService
import com.depromeet.threedollar.common.exception.model.InternalServerException
import com.depromeet.threedollar.domain.mongo.boss.domain.account.BossAccountSocialType
import org.springframework.stereotype.Component
import java.util.*
import javax.annotation.PostConstruct

private val authServiceMap: MutableMap<BossAccountSocialType, AuthService> = EnumMap(BossAccountSocialType::class.java)

@Component
class AuthServiceProvider(
    private val kaKaoAuthService: KaKaoAuthService,
    private val appleAuthService: AppleAuthService,
    private val googleAuthService: GoogleAuthService,
    private val naverAuthService: NaverAuthService
) {

    @PostConstruct
    fun initializeAuthServicesMap() {
        authServiceMap[BossAccountSocialType.KAKAO] = kaKaoAuthService
        authServiceMap[BossAccountSocialType.APPLE] = appleAuthService
        authServiceMap[BossAccountSocialType.GOOGLE] = googleAuthService
        authServiceMap[BossAccountSocialType.NAVER] = naverAuthService
    }

    fun getAuthService(socialType: BossAccountSocialType): AuthService {
        return authServiceMap[socialType] ?: throw InternalServerException("AuthService ($socialType) 로직이 구현되지 않았습니다")
    }

}
